README.md
Overview
========

This cookbook contains recipes for installing and configuring OMERO.server 
and OMERO.web.

Prerequisites:
--------------

If you deploy using Chef Solo, you need to set 
`node['postgresql']['password']['postgres']` to a password for the database.  
(By default, the postgres cookbook wants to persist a randomly generated 
password in a node attribute ... which only works in Chef Server mode.)  
Refer to the postgres cookbook's README file for more details, and 
instructions on how to generate the password hash.

You need to open firewall access on ports 4063 and 4064 for
OMERO.service, and optionally ports 80 and 443 for Omero.web.  (If you 
use non-standard ports, adjust accordingly.)  

For a NeCTAR virtual, you need to open the external firewall ports using
the NeCTAR Dashboard:
1. Go to your project's "Access & Security" panel
2. Click "Edit Rules".
3. Click "Add Rule", fill in the port and then "Add".  
4. Repeat for each port you need to open.

Recipes:
--------

The `omero::server` recipe installs and configures OMERO.server according 
to the standard OMERO instructions.

The `omero::web` recipe runs the server recipe, and then installs and 
configures OMERO.web using Nginx as the front-end web-server.

The `omero::web-internal` recipe runs the server recipe, and then installs and 
configures Omero.web in development mode.  (The omero-web service is disabled,
and you need to run it by hand using ".../OMERO.server/bin/omero web start".)

Attributes:
-----------

The following attributes control the main OMERO installation recipe:

* `node['omero']['release']` - The OMERO release to be installed.  This can be either the URL for the OMERO.server downloadable, or its basename.  Note that this recipe only supports 5.x releases.
* `node['omero']['install']` - The OMERO installation directory; defaults to 
"/opt/omero".  The installation will be created as a subdirectory of this,
and an "OMERO.server" symlink will be created to the installation.
* `node['omero']['user']` - The OMERO service user name; defaults to "omero".
* `node['omero']['var']` - The OMERO server's 'var' directory; defaults to 
"/var/omero".
* `node['omero']['db_user']` - The OMERO server's postgres database account 
name; defaults to "db_user".
* `node['omero']['db_password']` - The OMERO server's postgres database account 
password; defaults to "db_password".  THIS SHOULD BE CHANGED.
* `node['omero']['db_name']` - The OMERO server's postgres database name; 
defaults to "omero_database".
* `node['omero']['root_password']` - The OMERO root password; defaults to 
"omero_root_password".  THIS SHOULD BE CHANGED.

The following attributes control the OMERO web recipes:

* `node['omero']['web']['enabled']` - This tells the `omero::web` recipe to
enable or disable the `omero-web` init script; defaults to true.
* `node['omero']['web']['frontend']` - Select the frontend webserver.  
Currently, only 'nginx' is supported.
* `node['omero']['web']['http_port']` - Select frontend webserver's HTTP port;  
defaults to 80.
* `node['omero']['web']['recipe']` - Select the Chef recipe to use to install
the frontend webserver.  By default, we use the "standard" Chef Community 
cookbook.
* `node['omero']['web']['configure']` - If true, the web recipe will attempt 
to configure the frontend webserver on the assumption that it only needs to 
support Omero.web.  If false, we don't touch the frontend configuration.  
Defaults to true. 

Starting and stopping services
------------------------------

The recipe installs service scripts in "/etc/init.d" so that you can start
and stop the services.  The "omero" script is for the OMERO.service services,
and the "omero-web" script is for the OMERO.web services.

For example:

```
  $ sudo /etc/init.d/omero start      # starts the OMERO service
  $ sudo /etc/init.d/omero stop       # stops the OMERO service
  $ sudo /etc/init.d/omero restart    # stops and starts the OMERO service
```

Testing
-------

If the OMERO.server installation is working, you should be use OMERO.client
tools to talk to the server on (at least) port 4063 of the server, using the
server's external IP address or hostname.  (The OMERO.client tools can be
downloaded and installed on Windows, Mac and Linux machines.)

If the OMERO.web installation is working, you should be able to access the
OMERO web service using a web browser.  Use "http://<IP-address>/" as the URL,
where <IP> is the IP address of the server.  (If you have configured the server
to use a non-standard port, include that in the URL.)

In both cases, you can use the "root" account and password for initial testing.

More documentation
------------------

For more documentation, please refer to the relevant OMERO documention 
pages on their support site:

* http://www.openmicroscopy.org/site/support/omero5/

The installation documentation has various hints and links on checking
things are working, and diagnosing problems:

* http://www.openmicroscopy.org/site/support/omero5/sysadmins/unix/server-installation.html

Limitations
-----------

The cookbook only works on Ubuntu 14.04 and Fedora 19/20 at the moment, and 
only supports 'nginx' as the webserver front-end.

I have tried to get 'apache' working on Ubuntu 14.04, but at this time the
community cookbook for Apache2 does not support Apache 2.4.  Since Apache 2.4 
is the standard for Ubuntu 13.10 and Fedora 19+, that means the 'apache' 
won't work there either.

Fedora support requires a fix to the community "postgresql" cookbook.  I've
sent a pull request, but in the meantime you can get hold of the fixes by
adding this line to your Berksfile, updating and re-vendoring.

```
cookbook 'postgresql', github: 'crawley/postgresql', branch: 'fedora16+fixes'

```

TO DO List
----------

* We are not yet doing anything about backup or internal firewalls.
* We are not attempting to configure advanced options such as LDAP, 
  OMERO.dropbox, though apparently OMERO.dropbox is running ...
* Support for Omero.web needs more work / testing
* Support more OS platforms
* Support for SSL i.e. generating and configuring a certificate
* Support for updating Omero
* Support for rerunning with different usernames, passwords, etc.
* Support for installing clients
* Support for apache frontends (depending on community cookbook updates).
* Support for updating postgresql(?).


ChangeLog.md
Change Log for the Omero cookbook
=================================

Version 0.5.5
-------------
 - Support for CentOS 6.5 / SL 6.5 / RHEL 6.5 (tested on CentOS 6.5)

Version 0.5.4
-------------
 - Support for Fedora (tested on Fedora 20), but see limitations.

Version 0.5.3
-------------
 - Use fastcgi-tcp instead of fastcgi (Issue #5)

Version 0.5.2
-------------
 - First working version of "omero::web" recipe.

Version 0.5.1
-------------
 - It is no longer necessary to hack etc/grid/default.xml to get OMERO
   to listen the external IP address.  (Issue #3)

Version 0.5
-----------
 - First beta release.


recipes/web.rb
#
# Cookbook Name:: omero
# Recipe:: web
#
# Copyright (c) 2014, The University of Queensland
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# * Redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer.
# * Redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution.
# * Neither the name of the The University of Queensland nor the
# names of its contributors may be used to endorse or promote products
# derived from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE UNIVERSITY OF QUEENSLAND BE LIABLE FOR
# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

include_recipe 'omero::server'

omero_install = node['omero']['install']
omero_user = node['omero']['user']
omero_var = node['omero']['var']
enabled = node['omero']['web']['enabled']
web_frontend = node['omero']['web']['frontend']
web_port = node['omero']['web']['http_port']
web_configure = node['omero']['web']['configure']
web_recipe = node['omero']['web']['recipe']

# If the server might be running, stop and disable it.
service 'omero-web stop' do
  service_name 'omero-web'
  pattern 'OMERO.server/var/django.pid'
  action [ :stop ] 
  provider Chef::Provider::Service::Init
  only_if do ::File.exists?('/etc/init.d/omero-web') end
end

bash 'omero-web-configuration' do
  cwd "#{omero_install}/OMERO.server"
  user omero_user
  code <<-EOH
    bin/omero config set omero.web.application_server "fastcgi-tcp"
    bin/omero config set omero.web.debug False
    bin/omero config set omero.web.server_list \
            '[["#{node['ipaddress']}", 4064, "omero"]]'
    EOH
end

if web_frontend then
  if web_frontend == 'apache' then
    web_recipe ||= 'apache2::default'
    ensite = 'a2ensite'
    dissite = 'a2dissite'
    http = node['apache']['dir']
  elsif web_frontend == 'nginx' then
    web_recipe ||= 'nginx::default'     
    ensite = 'nxensite'
    dissite = 'nxdissite'
    http = node['nginx']['dir']
    if platform_family?('rhel') 
      include_recipe 'nginx::repo'
    end
  else 
    raise "Unsupported web frontend #{web_frontend}"
  end
  include_recipe web_recipe
  if web_configure then
    web_opts = "--system --http #{web_port}"
    bash 'omero-web-frontend' do
      cwd "#{omero_install}/OMERO.server"
      code <<-EOH
        sudo -u #{omero_user} bash -c "export HOME=/tmp ; bin/omero web config #{web_frontend} #{web_opts}" \
                > #{http}/sites-available/omero
        #{dissite} default
        #{ensite} omero
        # Disable the "default" config on nginx if it exists
        if [ -e #{http}/conf.d/default.conf ] ; then
          mv #{http}/conf.d/default.conf #{http}/conf.d/default.conf.xxx
        fi
      EOH
    end
  end
end


if platform_family?('debian') then
  rc_flavour = '-debian'
else
  rc_flavour = ''
end

template '/etc/init.d/omero-web' do
  source "omero-web-init#{rc_flavour}.erb"
  mode 0755
  variables({
     :omero_user => omero_user,
     :omero_home => "#{omero_install}/OMERO.server"
  })            
end

if enabled then
  service 'omero-web start' do
    service_name 'omero-web'
    pattern 'OMERO.server/var/django.pid'
    provider Chef::Provider::Service::Init
    action :start
  end
  
  service 'omero-web enable' do
    service_name 'omero-web'
    pattern 'OMERO.server/var/django.pid'
    action :enable
  end
else 
  service 'omero-web disable' do
    service_name 'omero-web'
    pattern 'OMERO.server/var/django.pid'
    action :disable
  end
end


attributes/default.rb
node.default['omero']['release'] = 
  "http://downloads.openmicroscopy.org/omero/5.0.2/artifacts/OMERO.server-5.0.2-ice35-b26.zip"
node.default['omero']['install'] = '/opt/omero'
node.default['omero']['user'] = 'omero'
node.default['omero']['var'] = '/var/omero'
node.default['omero']['db_user'] = 'db_user'
node.default['omero']['db_password'] = 'db_password'
node.default['omero']['db_name'] = 'omero_database'
node.default['omero']['root_password'] = 'omero_root_password'

# The 'web' recipes use this to determine whether the Omero web service
# is enabled / started or disabled / stopped by the recipe.
node.default['omero']['web']['enabled'] = true

# The supported choices are 'nginx' or 'apache'
node.default['omero']['web']['frontend'] = 'nginx'

node.default['omero']['web']['http_port'] = 80

# The recipe used to install and configure the frontend webserver.  If unset,
# the 'web.rb' recipe intuits one based on the 'frontend' attribute.
node.default['omero']['web']['recipe'] = nil

# If this is true, we attempt to inject suitable reverse proxy configs into
# the frontend webserver configurations.
node.default['omero']['web']['configure'] = true

node.normal['java']['install_flavor'] = 'openjdk'
node.normal['java']['jdk_version'] = '7'
node.normal['java']['accept_license_agreement'] = true


recipes/web-internal.rb
#
# Cookbook Name:: omero
# Recipe:: web-internal
#
# Copyright (c) 2014, The University of Queensland
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# * Redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer.
# * Redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution.
# * Neither the name of the The University of Queensland nor the
# names of its contributors may be used to endorse or promote products
# derived from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE UNIVERSITY OF QUEENSLAND BE LIABLE FOR
# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

include_recipe 'omero::server'

omero_install = node['omero']['install']
omero_user = node['omero']['user']
omero_var = node['omero']['var']

# If the server might be running, stop and disable it.
service 'omero-web' do
  pattern 'OMERO.server/var/django.pid'
  action [ :disable, :stop ] 
  only_if do ::File.exists?('/etc/init.d/omero-web') end
end

bash 'omero-web-internal-configuration' do
  cwd "#{omero_install}/OMERO.server"
  user omero_user
  code <<-EOH
  bin/omero config set omero.web.application_server development
  bin/omero config set omero.web.debug True
  EOH
end



recipes/server.rb
#
# Cookbook Name:: omero
# Recipe:: server
#
# Copyright (c) 2014, The University of Queensland
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# * Redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer.
# * Redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution.
# * Neither the name of the The University of Queensland nor the
# names of its contributors may be used to endorse or promote products
# derived from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE UNIVERSITY OF QUEENSLAND BE LIABLE FOR
# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

omero_install = node['omero']['install']
omero_user = node['omero']['user']
omero_var = node['omero']['var']
db_user = node['omero']['db_user']
db_password = node['omero']['db_password']
db_name = node['omero']['db_name']
root_password = node['omero']['root_password']

BASE_URL = 'http://downloads.openmicroscopy.org/omero'
release = node['omero']['release']

unless %r{^((http|https)://.+/)?([^/]*)\.zip$} =~ release then
  raise "The release attribute should be a ZIP filename or url."
end
m1 = Regexp.last_match
build = m1[3]
unless %r{^[^-]*-([0-9.]+)(-ice[0-9]+)-(b[0-9]+)$} =~ build then
  raise "Cannot parse the release filename."
end
m2 = Regexp.last_match
version = m2[1]
ice = m2[2]
if m1[1] then
  url = release
else
  url = "#{BASE_URL}/#{version}/artifacts/#{build}.zip"
end

pips = []
if platform_family?('debian') then
  dependencies = [ 'zip', 'python2.7', 'python-matplotlib',
                   'python-numpy', 'python-scipy', 'python_tables',
                   "zeroc#{ice}", 'mencoder', 'postgresql' ]
  
  if platform?('ubuntu') && ( node['platform_version'] <=> '14.04') >= 0
    dependencies << 'python-pil'
  else
    pips = [ { 'module' => 'pil',
               'deps' => [ 'python-dev', 'libjpeg-dev', 'libfreetype6-dev', 
                           'zlib1g-dev' ]
             } ]
  end
    
elsif platform_family?('fedora') then
  dependencies = [ 'zip', 'unzip', 'python', 'python-devel', 'python_tables',
                   'python-pillow', 'python-matplotlib', 'numpy', 'scipy',
                   'ice', 'ice-python', 'ice-servers', 'mencoder', 
                   'postgresql-server' ]
  enable_rpmfusion_free = true

elsif platform_family?('rhel') then
  # Note that we are not installing any version of mencoder ...
  dependencies = [ 'zip', 'unzip', 'python', 'python-devel', 
                   'python-matplotlib', 'numpy', 'scipy',
                   'ice', 'ice-python', 'ice-servers' ]
  pips = [ { 'module' => 'pillow',
             'deps' => [ 'python-devel', 'libjpeg-devel', 'libtiff-devel', 
                         'zlib-devel' ]
           },
           { 'module' => 'numexpr',
             'version' => '1.4.2',
             'deps' => [],
           },
           { 'module' => 'tables',
             'version' => '2.4.0',
             'deps' => [ 'gcc', 'Cython', 'hdf5-devel' ]
           }
         ]
  enable_rpmfusion_free = true
  
  # Ugly way to add the ZeroC repo ...
  remote_file '/etc/yum.repos.d/zeroc-ice-el6.repo' do
    source 'http://download.zeroc.com/Ice/3.5/el6/zeroc-ice-el6.repo'
    not_if { ::File.exists?('/etc/yum.repos.d/zeroc-ice-el6.repo') }
  end

else
  raise "Platform / family not supported: #{node['platform']} / #{node['platform_family']}"
end

if enable_rpmfusion_free then
  # Fedora specific ...
  bash "enable rpmfusion" do
    code <<-EOF
    yum -y localinstall --nogpgcheck \
      http://download1.rpmfusion.org/free/fedora/rpmfusion-free-release-$(rpm -E %fedora).noarch.rpm 
    EOF
  end
end

# For any platforms that have an old version of postgresql ...
if platform_family?('rhel') && node['platform_version'].to_f() < 7.0 then
  node.normal['postgresql']['version'] = '9.3'
  node.normal['postgresql']['enable_pgdg_yum'] = true
  node.normal['postgresql']['client']['packages'] = ["postgresql93"]
  node.normal['postgresql']['server']['packages'] = ["postgresql93-server"]
  node.normal['postgresql']['server']['service_name'] = "postgresql-9.3"
end

include_recipe 'postgresql::client'
include_recipe 'postgresql::server'

include_recipe 'java::default'

# Regular packages from the package repositories
dependencies.each() do |pkg| 
  package pkg
end

# Python packages that need to be 'pip' installed.
if pips.length > 0
  include_recipe 'python::default'
  pips.each() do |pip_data|
    pip_data['deps'].each() do |pkg|
      package pkg do
      end
    end
    python_pip pip_data['module'] do
      version pip_data['version'] if pip_data['version']
    end
  end
end

user omero_user do
  comment 'Omero service user'
  home omero_var
  system true
  shell '/bin/false'
  supports :manage_home => true
end

directory omero_install do
  owner omero_user
end

remote_file "#{omero_install}/#{build}.zip" do
  source url
  action :create_if_missing
end

# Shut down Omero / Omero.web while we fiddle with things ...
service 'omero-web stop' do
  pattern 'OMERO.server/var/django.pid'
  service_name 'omero-web'
  provider Chef::Provider::Service::Init
  action [ :stop ] 
  only_if do ::File.exists?('/etc/init.d/omero-web') end
end
service 'omero stop' do
  service_name 'omero'
  pattern 'icegridnode'
  provider Chef::Provider::Service::Init
  action [ :stop ] 
  only_if do ::File.exists?('/etc/init.d/omero') end
end

bash "unpack" do
  code "unzip #{build}.zip"
  cwd omero_install
  user omero_user
  not_if do ::File.exists?("#{omero_install}/#{build}") end
end

link "#{omero_install}/OMERO.server" do
  to "#{omero_install}/#{build}"
  link_type :symbolic
end

bash 'init-postgres-database' do
  user 'postgres'
  code <<-EOH
  psql -d postgres -c "create user #{db_user} with password '#{db_password}';"
  createdb -O #{db_user} #{db_name}
  # createlang plpgsql #{db_name}
  EOH
  not_if <<-EOH
  export PGPASSWORD=#{db_password}
  psql -h localhost -U #{db_user} -l | grep -l #{db_name}
  EOH
end

directory "#{omero_var}/data" do
  user omero_user
  mode 0750
end

# (For reasons that we don't understand, it is necessary to tell Ice
# to use the virtual's IP address as the default host.  This ought to
# happen without us saying so, but ...)
bash 'omero-configuration' do
  cwd "#{omero_install}/OMERO.server"
  user omero_user
  code <<-EOH
  bin/omero config set omero.db.name #{db_name}
  bin/omero config set omero.db.user #{db_user}
  bin/omero config set omero.db.pass #{db_password}
  bin/omero config set omero.data.dir #{omero_var}/data
  bin/omero config set Ice.Default.Host #{node['ipaddress']}
  EOH
end

# 
bash 'init-omero-db' do
  cwd "#{omero_install}/OMERO.server"
  user omero_user
  code <<-EOH
  export HOME=/tmp
  bin/omero db script "" "" #{root_password}
  export PGPASSWORD=#{db_password}
  psql -h localhost -U #{db_user} #{db_name} < OMERO5.0__0.sql
  EOH
  not_if do ::File.exists?("#{omero_install}/OMERO.server/OMERO5.0__0.sql") end
end

if platform_family?('debian') then
  rc_flavour = '-debian'
else
  rc_flavour = ''
end

template '/etc/init.d/omero' do
  source "omero-init#{rc_flavour}.erb"
  mode 0755
  variables({
     :omero_user => omero_user,
     :omero_home => "#{omero_install}/OMERO.server"
  })            
end

service 'omero start' do
  service_name 'omero'
  pattern 'icegridnode'
  provider Chef::Provider::Service::Init
  action [ :start ]
end

service 'omero enable' do
  service_name 'omero'
  pattern 'icegridnode'
  action [ :enable ]
end
